File: desktop/Dms.Desktop.DriverHud/Services/IncidentQueueService.cs
using System.IO;
using System.Text.Json;
using Dms.Client.Api;
using Dms.Shared.Contracts.Telemetry;
using Microsoft.Data.Sqlite;
using Microsoft.Extensions.Logging;

namespace Dms.Desktop.DriverHud.Services;

/// <summary>
/// Offline incident queue: an incident is always written to a local SQLite
/// db first, then a background loop tries to POST each queued row to
/// TelemetryController and deletes it on success. If the device is in a
/// cellular/network dead zone, incidents just accumulate locally until the
/// next successful sync — same contract as the mobile app's Room queue.
/// </summary>
public class IncidentQueueService : IAsyncDisposable
{
    private readonly DmsApiClient _api;
    private readonly ILogger<IncidentQueueService> _logger;
    private readonly string _dbPath;
    private readonly string _connectionString;
    private CancellationTokenSource? _cts;
    private Task? _syncLoop;

    public event Action<int>? PendingCountChanged;
    public event Action<IncidentReport>? IncidentSynced;

    public IncidentQueueService(DmsApiClient api, ILogger<IncidentQueueService> logger)
    {
        _api = api;
        _logger = logger;

        var directory = Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "AwakeDrive");
        Directory.CreateDirectory(directory);
        _dbPath = Path.Combine(directory, "incident-queue.db");
        _connectionString = $"Data Source={_dbPath}";

        InitializeDatabase();
    }

    private void InitializeDatabase()
    {
        using var connection = new SqliteConnection(_connectionString);
        connection.Open();
        using var command = connection.CreateCommand();
        command.CommandText = """
            CREATE TABLE IF NOT EXISTS PendingIncidents (
                Id INTEGER PRIMARY KEY AUTOINCREMENT,
                PayloadJson TEXT NOT NULL,
                CreatedAtUtc TEXT NOT NULL
            );
            """;
        command.ExecuteNonQuery();
    }

    public async Task EnqueueAsync(IncidentReport report)
    {
        await using var connection = new SqliteConnection(_connectionString);
        await connection.OpenAsync();
        await using var command = connection.CreateCommand();
        command.CommandText = "INSERT INTO PendingIncidents (PayloadJson, CreatedAtUtc) VALUES ($payload, $createdAt);";
        command.Parameters.AddWithValue("$payload", JsonSerializer.Serialize(report));
        command.Parameters.AddWithValue("$createdAt", DateTimeOffset.UtcNow.ToString("O"));
        await command.ExecuteNonQueryAsync();

        PendingCountChanged?.Invoke(await GetPendingCountAsync());
    }

    public async Task<int> GetPendingCountAsync()
    {
        await using var connection = new SqliteConnection(_connectionString);
        await connection.OpenAsync();
        await using var command = connection.CreateCommand();
        command.CommandText = "SELECT COUNT(*) FROM PendingIncidents;";
        return Convert.ToInt32(await command.ExecuteScalarAsync());
    }

    public void StartBackgroundSync(TimeSpan? interval = null)
    {
        if (_syncLoop is not null)
        {
            return;
        }
        _cts = new CancellationTokenSource();
        var syncInterval = interval ?? TimeSpan.FromSeconds(15);
        _syncLoop = Task.Run(() => SyncLoop(syncInterval, _cts.Token));
    }

    private async Task SyncLoop(TimeSpan interval, CancellationToken ct)
    {
        while (!ct.IsCancellationRequested)
        {
            try
            {
                await FlushOnceAsync(ct);
            }
            catch (Exception ex)
            {
                _logger.LogWarning(ex, "Incident queue flush failed; will retry");
            }

            try
            {
                await Task.Delay(interval, ct);
            }
            catch (OperationCanceledException)
            {
                break;
            }
        }
    }

    private async Task FlushOnceAsync(CancellationToken ct)
    {
        await using var connection = new SqliteConnection(_connectionString);
        await connection.OpenAsync(ct);

        await using var selectCommand = connection.CreateCommand();
        selectCommand.CommandText = "SELECT Id, PayloadJson FROM PendingIncidents ORDER BY Id ASC LIMIT 20;";

        var rows = new List<(long Id, string Payload)>();
        await using (var reader = await selectCommand.ExecuteReaderAsync(ct))
        {
            while (await reader.ReadAsync(ct))
            {
                rows.Add((reader.GetInt64(0), reader.GetString(1)));
            }
        }

        foreach (var (id, payloadJson) in rows)
        {
            var report = JsonSerializer.Deserialize<IncidentReport>(payloadJson);
            if (report is null)
            {
                continue; // corrupt row — drop it rather than blocking the queue forever
            }

            try
            {
                await _api.PostIncidentAsync(report, ct);
            }
            catch (DmsApiException)
            {
                // Backend unreachable or rejected it — stop this pass, leave
                // remaining rows queued, try again next interval.
                return;
            }

            await using var deleteCommand = connection.CreateCommand();
            deleteCommand.CommandText = "DELETE FROM PendingIncidents WHERE Id = $id;";
            deleteCommand.Parameters.AddWithValue("$id", id);
            await deleteCommand.ExecuteNonQueryAsync(ct);

            IncidentSynced?.Invoke(report);
            PendingCountChanged?.Invoke(await GetPendingCountAsync());
        }
    }

    public async ValueTask DisposeAsync()
    {
        _cts?.Cancel();
        if (_syncLoop is not null)
        {
            try
            {
                await _syncLoop;
            }
            catch (OperationCanceledException)
            {
                // expected on shutdown
            }
        }
        _cts?.Dispose();
    }
}

File: mobile/awakedrive_driver/lib/services/incident_queue_service.dart
import 'dart:async';
import 'dart:convert';
import 'package:path/path.dart';
import 'package:sqflite/sqflite.dart';

import '../models/contracts.dart';
import 'api_client.dart';

/// Offline incident queue: an incident is always written to a local SQLite
/// db first, then a background timer tries to POST each queued row to
/// TelemetryController and deletes it on success. If the phone is in a
/// cellular dead zone, incidents just accumulate locally until the next
/// successful sync — the Dart equivalent of the desktop HUD's
/// IncidentQueueService (Room-queue-equivalent contract from the spec).
class IncidentQueueService {
  final DmsApiClient api;
  Database? _db;
  Timer? _syncTimer;

  final _pendingCountController = StreamController<int>.broadcast();
  Stream<int> get pendingCountStream => _pendingCountController.stream;

  IncidentQueueService({required this.api});

  Future<Database> _database() async {
    if (_db != null) return _db!;
    final path = join(await getDatabasesPath(), 'incident_queue.db');
    _db = await openDatabase(
      path,
      version: 1,
      onCreate: (db, version) => db.execute('''
        CREATE TABLE PendingIncidents (
          id INTEGER PRIMARY KEY AUTOINCREMENT,
          payloadJson TEXT NOT NULL,
          createdAtUtc TEXT NOT NULL
        )
      '''),
    );
    return _db!;
  }

  Future<void> enqueue(IncidentReport report) async {
    final db = await _database();
    await db.insert('PendingIncidents', {
      'payloadJson': jsonEncode(report.toJson()),
      'createdAtUtc': DateTime.now().toUtc().toIso8601String(),
    });
    _pendingCountController.add(await pendingCount());
  }

  Future<int> pendingCount() async {
    final db = await _database();
    final result = await db.rawQuery('SELECT COUNT(*) AS c FROM PendingIncidents');
    return Sqflite.firstIntValue(result) ?? 0;
  }

  void startBackgroundSync({Duration interval = const Duration(seconds: 15)}) {
    _syncTimer?.cancel();
    _syncTimer = Timer.periodic(interval, (_) => flushOnce());
    // Also try once immediately rather than waiting for the first tick.
    flushOnce();
  }

  void stopBackgroundSync() {
    _syncTimer?.cancel();
    _syncTimer = null;
  }

  Future<void> flushOnce() async {
    final db = await _database();
    final rows = await db.query('PendingIncidents', orderBy: 'id ASC', limit: 20);

    for (final row in rows) {
      final report = IncidentReport.fromJson(jsonDecode(row['payloadJson'] as String) as Map<String, dynamic>);
      try {
        await api.postIncident(report);
      } on DmsApiException {
        // Backend unreachable or rejected it — stop this pass, leave
        // remaining rows queued, try again next interval.
        return;
      }
      await db.delete('PendingIncidents', where: 'id = ?', whereArgs: [row['id']]);
      _pendingCountController.add(await pendingCount());
    }
  }

  void dispose() {
    _syncTimer?.cancel();
    _pendingCountController.close();
  }
}
